### ISSUE TITLE
[Beta] Noticeable delay in autocomplete suggestions showing up after insertion

### ISSUE THREAD

**Issue by brendanfalk** [withfig/fig#923]
After updating to the most recent beta (build 374), I started to notice the suggestions popped up slower than before. This was especially in the cd suggestions after inserting the first directory. e.g. going from `cd[space]` to `cd desktop/`

According to the videos, the time between the text being inserted and the popup showing has almost doubled (from ~60ms to ~110ms). Unfortunately, these statistics seems to be accurate in the +/- 33ms range (as the screencast was 30FPS) AND the millisecond stopwatch wasn't running smoothly (as in it didn't directly count from 1.001 to 1.002, rather 1.001 to around 1.035).

In any case, the difference is almost 2 lots of 33ms and I believe that this is therefore significant enough to be true. 

Below, I compare the videos as if it's a c figterm vs rust figterm issue. Disregard this. Just think of it as prod build vs beta build.

### Comparison Vid (between rust and C figterm)

https://user-images.githubusercontent.com/4949076/153340596-ec581c6b-b9bc-4d3d-8df1-2f256ceb6a62.mp4

### Original vid for C

https://user-images.githubusercontent.com/4949076/153340664-9b286893-4d32-41f8-8731-b9dac3f1be84.mp4

### Original vid for rust

https://user-images.githubusercontent.com/4949076/153340711-678d7fc6-27d5-4991-9ed4-dc9c66086d31.mp4

### Original vid for website

https://user-images.githubusercontent.com/4949076/153340718-7fa2a1db-d138-4237-8fac-4bc9ee7eaf42.mp4

Optimal outcome here is to get it so suggestions appears as fast as possible. Remember, we are competing with shell suggestions which are lightning fast.  

In the past we have had users leave as Fig was popping up slowly. This was really early days Fig. I don't believe I have seen users leave because suggestions have popped up too slowly recently (ie in past few months), except for users who had bugs.

I understand there are engineering constraints that may prevent us from getting back down to 30ms. I would love to do this one day. However, I think we should at least get back to levels we were experiencing with c figterm. 

Matt seems to know what the cause is (comment below). It does not seem to be anything to do with figterm, regardless, thought it would be worth mentioned that this is the beta where we replaced the c version of figterm with the rust version of figterm.

**Comment by mschrage** [comment]
This is not actually a Rust issue. 

Sean and I removed a delay around the insertion lock in the previous version that we thought was no longer necessary. Getting rid of the delay introduced a bug (folder suggestions not appearing after an insertion). We reintroduced the delay in the latest version to fix this issue. It has nothing to do with the figterm rewrite.

We have an idea for how to fix this. Sean mentioned it yesterday when you brought this up. It has to do with refactoring how we do insertions, so that figterm  can send the edit buffer update as an atomic event rather than using a delay-based insertion lock and relying on the macOS app to send the event.
